frappe/email/doctype/email_alert/email_alert.py

```diff
@@ -171,13 +142,55 @@ def get_attachment(doc):
 
 		attachments = get_attachment(doc)
 
-		frappe.sendmail(recipients=recipients, subject=subject,
-			message= frappe.render_template(self.message, context),
-			reference_doctype = doc.doctype,
-			reference_name = doc.name,
-			attachments = attachments,
-			print_letterhead = ((attachments
-				and attachments[0].get('print_letterhead')) or False))
+		if self.channel == 'Email':
+			recipients = []
+			for recipient in self.recipients:
+				if recipient.condition:
+					if not frappe.safe_eval(recipient.condition, None, context):
+						continue
+				if recipient.email_by_document_field:
+					if validate_email_add(doc.get(recipient.email_by_document_field)):
+						recipient.email_by_document_field = doc.get(recipient.email_by_document_field).replace(",", "\n")
+						recipients = recipients + recipient.email_by_document_field.split("\n")
+
+					# else:
+					# 	print "invalid email"
+				if recipient.cc and "{" in recipient.cc:
+					recipient.cc = frappe.render_template(recipient.cc, context)
+
+				if recipient.cc:
+					recipient.cc = recipient.cc.replace(",", "\n")
+					recipients = recipients + recipient.cc.split("\n")
+
+				#For sending emails to specified role
+				if recipient.email_by_role:
+					emails = get_emails_from_role(recipient.email_by_role)
+
+					for email in emails:
+						recipients = recipients + email.split("\n")
+
+			if not recipients:
+				return
+
+			recipients = list(set(recipients))
+
+			frappe.sendmail(recipients=recipients, subject=subject,
+				message= frappe.render_template(self.message, context),
+				reference_doctype = doc.doctype,
+				reference_name = doc.name,
+				attachments = attachments,
+				print_letterhead = ((attachments
+					and attachments[0].get('print_letterhead')) or False))
+
+		elif self.channel == 'Slack':
```

Comment:
Make smaller functions. This function is too long.

Reply:
Sure I'll refactor it